src/backend/gl/src/window/surfman.rs

```diff
@@ -0,0 +1,346 @@
+        })
+    }
+
+    fn enumerate_adapters(&self) -> Vec<Adapter<B>> {
+        let mut adapters = Vec::with_capacity(3);
+
+        let context_attributes = Self::get_default_context_attributes();
+
+        for surfman_adapter in &[
+            &self.hardware_adapter,
+            // TODO: Enabling these causes a segfault for an unknown reason
+            // &self.low_power_adapter,
+            // &self.software_adapter,
+        ] {
+            // Create a surfman device
+            let mut device =
+                SM_CONN.with(|c| c.borrow().create_device(surfman_adapter).expect("TODO"));
+
+            // Create context descriptor
+            let context_descriptor = device
+                .create_context_descriptor(&context_attributes)
+                .expect("TODO");
+
+            // Create context
+            let context = device.create_context(&context_descriptor).expect("TODO");
+
+            // Wrap in Starc<RwLock<T>>
+            let context = Starc::new(RwLock::new(context));
+            let context_ = context.clone();
+            let device = Starc::new(RwLock::new(device));
+            let device_ = device.clone();
+
+            // Create gl container
+            let gl = GlContainer::from_fn_proc(
+                |symbol_name| {
+                    device_
+                        .write()
+                        .get_proc_address(&context_.read(), symbol_name)
+                        as *const _
+                },
+                device,
+                context,
+            );
+
+            // Create physical device
+            adapters.push(PhysicalDevice::new_adapter((), gl));
+        }
+
+        adapters
+    }
+
+    unsafe fn create_surface(
+        &self,
+        has_handle: &impl raw_window_handle::HasRawWindowHandle,
+    ) -> Result<Surface, window::InitError> {
+        Ok(self.create_surface_from_rwh(has_handle.raw_window_handle()))
+    }
+
+    unsafe fn destroy_surface(&self, surface: Surface) {
+        // Surface implments Drop and will clean up the surface when it gets dropped
+        drop(surface);
+    }
+}
+
+// TODO: Not sure if this TODO is relevant with surfman.
+// TODO: if we make `Surface` a `WindowBuilder` instead of `RawContext`,
+// we could spawn window + GL context when a swapchain is requested
+// and actually respect the swapchain configuration provided by the user.
+#[derive(Debug)]
+pub struct Surface {
+    pub(crate) swapchain: Option<Swapchain>,
+    pub(crate) context: Starc<RwLock<sm::Context>>,
+    device: Starc<RwLock<sm::Device>>,
+    renderbuffer: Option<native::Renderbuffer>,
+}
+
+impl Surface {
+    pub fn context(&self) -> Starc<RwLock<sm::Context>> {
+        self.context.clone()
+    }
+
+    fn swapchain_formats(&self) -> Vec<f::Format> {
+        // TODO: Make sure this is correct. I believe it is. Reference:
+        // http://docs.rs/surfman/struct.ContextAttributeFlags.html#associatedconstant.ALPHA
```

Comment:
that link is dead

Reply:
Ah, yeah, the docs I was referring to aren't released yet ( and I botched the link, too ).

I'll link to the doc string on GitHub for now.